Source/FieldSolver/MagnetostaticSolver/MagnetostaticSolver.cpp

```diff
@@ -0,0 +1,462 @@
+
+#include <ablastr/utils/Communication.H>
+#include <ablastr/warn_manager/WarnManager.H>
+#include <ablastr/fields/VectorPoissonSolver.H>
+
+#include <AMReX_Array.H>
+#include <AMReX_Array4.H>
+#include <AMReX_BLassert.H>
+#include <AMReX_Box.H>
+#include <AMReX_BoxArray.H>
+#include <AMReX_Config.H>
+#include <AMReX_DistributionMapping.H>
+#include <AMReX_FArrayBox.H>
+#include <AMReX_FabArray.H>
+#include <AMReX_Geometry.H>
+#include <AMReX_GpuControl.H>
+#include <AMReX_GpuLaunch.H>
+#include <AMReX_GpuQualifiers.H>
+#include <AMReX_IndexType.H>
+#include <AMReX_IntVect.H>
+#include <AMReX_LO_BCTYPES.H>
+#include <AMReX_MFIter.H>
+#include <AMReX_MLMG.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_REAL.H>
+#include <AMReX_SPACE.H>
+#include <AMReX_Vector.H>
+#include <AMReX_MFInterp_C.H>
+#ifdef AMREX_USE_EB
+#   include <AMReX_EBFabFactory.H>
+#endif
+
+#include <array>
+#include <memory>
+#include <string>
+
+using namespace amrex;
+
+void
+WarpX::ComputeCurrentDensityField()
+{
+    WARPX_PROFILE("WarpX::ComputeCurrentDensityField");
+    // Fields have been reset in Electrostatic solver for this time step, these fields
+    // are added into the E & B fields after electrostatic solve
+
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(this->max_level == 0, "Magnetostatic solver not implemented with mesh refinement.");
+
+    AddCurrentDensityFieldLabFrame();
+}
+
+void
+WarpX::AddCurrentDensityFieldLabFrame()
+{
+    WARPX_PROFILE("WarpX::AddCurrentDensityFieldLabFrame");
+
+    // Store the boundary conditions for the field solver if they haven't been
+    // stored yet
+    if (!m_vector_poisson_boundary_handler.bcs_set) {
+        m_vector_poisson_boundary_handler.defineVectorPotentialBCs();
+    }
+
+#ifdef WARPX_DIM_RZ
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(n_rz_azimuthal_modes == 1,
+                                     "Error: RZ magnetostatic only implemented for a single mode");
+#endif
+
+    // reset current_fp before depositing current density for this step
+    for (int lev = 0; lev <= max_level; lev++) {
+        for (int dim=0; dim < 3; dim++) {
+            current_fp[lev][dim]->setVal(0.);
+        }
+    }
+
+    // Deposit current density (source of Poisson solver)
+    for (int ispecies=0; ispecies<mypc->nSpecies(); ispecies++){
+        WarpXParticleContainer& species = mypc->GetParticleContainer(ispecies);
+        if (!species.do_not_deposit) {
+            species.DepositCurrent(current_fp, dt[0], 0.);
+        }
+    }
+
+    // This will interpolate to staggered grid, filter, and synchronize
+    SyncCurrent(current_fp, current_cp); // Apply filter, perform MPI exchange, interpolate across levels
+
+#ifdef WARPX_DIM_RZ
+    for (int lev = 0; lev <= max_level; lev++) {
+        ApplyInverseVolumeScalingToCurrentDensity(current_fp[lev][0].get(),
+                                                  current_fp[lev][1].get(),
+                                                  current_fp[lev][2].get(), lev);
+    }
+#endif
```

Comment:
We could exchange the order of these 2 operations, in order to be more consistent with `AddSpaceChargeLabFrame`:
```suggestion
#ifdef WARPX_DIM_RZ
    for (int lev = 0; lev <= max_level; lev++) {
        ApplyInverseVolumeScalingToCurrentDensity(current_fp[lev][0].get(),
                                                  current_fp[lev][1].get(),
                                                  current_fp[lev][2].get(), lev);
    }
#endif
SyncCurrent(current_fp, current_cp); // Apply filter, perform MPI exchange, interpolate across levels
```